power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -622,21 +610,24 @@ class TapPositionOptimizerImpl<std::tuple<TransformerTypes...>, StateCalculator,
         initialize(regulator_order);
 
         if (auto result = iterate_with_fallback(state, regulator_order, method); strategy_ == OptimizerStrategy::any) {
-            return produce_output(state, std::move(result));
+            return produce_output(regulator_order, std::move(result));
         }
 
         // refine solution
         exploit_neighborhood(regulator_order);
-        return produce_output(state, iterate_with_fallback(state, regulator_order, method));
+        return produce_output(regulator_order, iterate_with_fallback(state, regulator_order, method));
     }
 
-    auto produce_output(State const& state, ResultType solver_output) const -> MathOutput<ResultType> {
+    auto produce_output(std::vector<std::vector<RegulatedTransformer>> const& regulator_order,
+                        ResultType solver_output) const -> MathOutput<ResultType> {
         TransformerTapPositionResult transformer_tap_positions;
 
-        // TODO(mgovers): only output the transformers that are regulated
-        (get_transformer_tap_positions<TransformerTypes, typename State::ComponentContainer>(state,
-                                                                                             transformer_tap_positions),
-         ...);
+        for (auto const& sub_order : regulator_order) {
+            for (auto const& regulator : sub_order) {
```

Comment:
Sooo a nested `for` then

Reply:
it is what it is. it's fine, i guess. it is definitely more readable than more variadic stuff (that is handled internally in the transformer regulator reference class)